lib/list_versions.go

```diff
@@ -1,142 +1,165 @@
-func GetTFList(mirrorURL string, preRelease bool) ([]string, error) {
-
-	result, error := GetTFURLBody(mirrorURL)
+//GetTFLatest :  Get the latest terraform version given the hashicorp url
+func GetTFLatest(mirrorURL string, preRelease bool) (Release, error) {
+	releases, error := GetTFReleases(mirrorURL, preRelease)
 	if error != nil {
-		return nil, error
-	}
-
-	var tfVersionList tfVersionList
-	var semver string
-	if preRelease == true {
-		// Getting versions from body; should return match /X.X.X-@/ where X is a number,@ is a word character between a-z or A-Z
-		semver = `\/(\d+\.\d+\.\d+)(-[a-zA-z]+\d*)?"`
-	} else if preRelease == false {
-		// Getting versions from body; should return match /X.X.X/ where X is a number
-		// without the ending '"' pre-release folders would be tried and break.
-		semver = `\/(\d+\.\d+\.\d+)\/?"`
-	}
-	r, _ := regexp.Compile(semver)
-	for i := range result {
-		if r.MatchString(result[i]) {
-			str := r.FindString(result[i])
-			trimstr := strings.Trim(str, "/\"") //remove '/' or '"' from /X.X.X/" or /X.X.X"
-			tfVersionList.tflist = append(tfVersionList.tflist, trimstr)
+		return Release{}, error
+	}
+	for i := range releases {
+		if !releases[i].IsPrerelease {
+			return releases[i], nil
 		}
 	}
+	return Release{}, nil
+}
 
-	if len(tfVersionList.tflist) == 0 {
-		fmt.Printf("Cannot get list from mirror: %s\n", mirrorURL)
+//GetTFLatestImplicit :  Get the latest implicit terraform version given the hashicorp url
+func GetTFLatestImplicit(mirrorURL string, preRelease bool, version string) (Release, error) {
+	if preRelease {
+		version = fmt.Sprintf(`%s{1}\.\d+\-[a-zA-z]+\d*`, version)
+	} else if !preRelease {
+		version = fmt.Sprintf("~> %v", version)
+	}
+	releases, err := GetTFReleases(mirrorURL, preRelease)
+	//version := fmt.Sprintf(`%s{1}\.\d+\-[a-zA-z]+\d*`, version)
+	semv, err := SemVerParser(&version, releases)
+	if err != nil {
+		return Release{}, err
 	}
+	return semv, nil
+}
 
-	return tfVersionList.tflist, nil
+func httpGet(url string, queryParams map[string]string) (*http.Response, error) {
+	req, err := http.NewRequest(http.MethodGet, url, nil)
+	if err != nil {
+		log.Fatal(err)
+	}
 
+	q := req.URL.Query()
+	for k, v := range queryParams {
+		q.Add(k, v)
+	}
+	req.URL.RawQuery = q.Encode()
+	client := &http.Client{}
+	res, err := client.Do(req)
+	if err != nil {
+		return nil, err
+	}
+	if res.StatusCode != http.StatusOK {
+		err := fmt.Errorf("issue during request (%d: %q)", res.StatusCode, res.Status)
+		return nil, err
+	}
+	return res, nil
 }
 
-//GetTFLatest :  Get the latest terraform version given the hashicorp url
-func GetTFLatest(mirrorURL string) (string, error) {
+func getReleases(mirrorURL string, queryParams map[string]string) ([]Release, error) {
+	var releases []Release
+	resp, errURL := httpGet(mirrorURL, queryParams)
+	if errURL != nil {
+		log.Printf("[Error] : Getting url: %v", errURL)
+		os.Exit(1)
```

Comment:
`log.Fatalf()`?
Or am I missing something essential?

Reply:
Looks like I missed a few places. Will take care of these as well.